src/node_url.cc

```diff
@@ -127,76 +76,160 @@ void DomainToASCII(const FunctionCallbackInfo<Value>& args) {
-#else
-  args.GetReturnValue().Set(
-      String::NewFromUtf8(env->isolate(), input.c_str()).ToLocalChecked());
-#endif
 }
 
-void DomainToUnicode(const FunctionCallbackInfo<Value>& args) {
+void BindingData::DomainToUnicode(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
   CHECK_GE(args.Length(), 1);
   CHECK(args[0]->IsString());
 
   std::string input = Utf8Value(env->isolate(), args[0]).ToString();
-#if defined(NODE_HAVE_I18N_SUPPORT)
   // It is important to have an initial value that contains a special scheme.
   // Since it will change the implementation of `set_hostname` according to URL
   // spec.
-  ada::result out = ada::parse("ws://x");
+  auto out = ada::parse<ada::url>("ws://x");
   DCHECK(out);
   if (!out->set_hostname(input)) {
     return args.GetReturnValue().Set(
         String::NewFromUtf8(env->isolate(), "").ToLocalChecked());
   }
-  std::string host = out->get_hostname();
+  std::string result = ada::unicode::to_unicode(out->get_hostname());
 
-  MaybeStackBuffer<char> buf;
-  int32_t len = i18n::ToUnicode(&buf, host.data(), host.length());
+  args.GetReturnValue().Set(String::NewFromUtf8(env->isolate(),
+                                                result.c_str(),
+                                                NewStringType::kNormal,
+                                                result.length())
+                                .ToLocalChecked());
+}
 
-  if (len < 0) {
-    return args.GetReturnValue().Set(
-        String::NewFromUtf8(env->isolate(), "").ToLocalChecked());
+// TODO(@anonrig): Add V8 Fast API for CanParse method
+void BindingData::CanParse(const FunctionCallbackInfo<Value>& args) {
+  CHECK_GE(args.Length(), 2);
+  CHECK(args[0]->IsString());  // input
+  // args[1] // base url
+
+  Environment* env = Environment::GetCurrent(args);
+  HandleScope handle_scope(env->isolate());
+  Context::Scope context_scope(env->context());
+
+  Utf8Value input(env->isolate(), args[0]);
+  ada::result<ada::url> base;
+  ada::url* base_pointer = nullptr;
+  if (args[1]->IsString()) {
+    base = ada::parse<ada::url>(Utf8Value(env->isolate(), args[1]).ToString());
+    if (!base) {
+      return args.GetReturnValue().Set(false);
+    }
+    base_pointer = &base.value();
+  }
+  auto out = ada::parse<ada::url>(input.ToStringView(), base_pointer);
+
+  args.GetReturnValue().Set(out.has_value());
+}
+
+void BindingData::Format(const FunctionCallbackInfo<Value>& args) {
+  CHECK_GT(args.Length(), 4);
+  CHECK(args[0]->IsString());  // url href
+
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+
+  Utf8Value href(isolate, args[0].As<String>());
+  const bool fragment = args[1]->IsTrue();
+  const bool unicode = args[2]->IsTrue();
+  const bool search = args[3]->IsTrue();
+  const bool auth = args[4]->IsTrue();
+
+  auto out = ada::parse<ada::url>(href.ToStringView());
```

Comment:
The code use a mix of `parse<ada::url>` and `parse<ada::url_aggregator>`. Both can be used in the same code base, evidently, but without further documentation, this could be puzzling to some.

Reply:
For this particular example, using `ada::url` is more performant due to having direct access to internal values. I'll add a comment to here to explain it.